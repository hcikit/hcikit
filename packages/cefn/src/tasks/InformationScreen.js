import React, { useEffect } from 'react'
import marked from 'marked'
import Typography from '@material-ui/core/Typography'
import Button from '@material-ui/core/Button'
import { CenteredNicePaper } from '../components'
import PropTypes from 'prop-types'
import { withGridItem } from '../core/App'

/**
 * Creates a screen with information or instructions and a continue button.
 * Accepts Markdown.
 * By setting withContinue to false it can also be used as an end screen of an experiment.
 * Uses a keyboard shortcut when specified.
 */
const InformationScreen = ({
  content,
  withContinue = true,
  onAdvanceWorkflow,
  centerX,
  centerY,
  shortcutEnabled,
  key = 'Enter'
}) => {
  useEffect(() => {
    function handleKeyDown(e) {
      if (e.key === key && shortcutEnabled) {
        onAdvanceWorkflow()
      }
    }
    window.addEventListener('keydown', handleKeyDown)
    return () => window.removeEventListener('keydown', handleKeyDown)
  })

  return (
    <CenteredNicePaper centerX={centerX} centerY={centerY}>
      <Typography dangerouslySetInnerHTML={{ __html: marked(content) }} />
      <br />
      {withContinue && (
        <Button color='primary' onClick={onAdvanceWorkflow}>
          continue {shortcutEnabled && `(${key})`}
        </Button>
      )}
    </CenteredNicePaper>
  )
}

InformationScreen.propTypes = {
  centerX: PropTypes.bool,
  centerY: PropTypes.bool,
  content: PropTypes.string,
  onAdvanceWorkflow: PropTypes.func,
  withContinue: PropTypes.bool,
  shortcutEnabled: PropTypes.bool,
  /* The KeyboardEvent.key value to use as a shortcut */
  key: PropTypes.string
}

export default withGridItem(InformationScreen)
