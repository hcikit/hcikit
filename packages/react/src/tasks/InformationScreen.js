import React, { useEffect } from "react";
import Markdown from "markdown-to-jsx";

import { Button } from "@material-ui/core";
import { CenteredNicePaper } from "../components";
import PropTypes from "prop-types";

/**
 * Creates a screen with information or instructions and a continue button.
 * Accepts Markdown.
 * By setting withContinue to false it can also be used as an end screen of an experiment.
 * Uses a keyboard shortcut when specified.
 */
const InformationScreen = ({
  content,
  withContinue = true,
  taskComplete,
  centerX,
  centerY,
  shortcutEnabled,
  key = "Enter",
}) => {
  useEffect(() => {
    function handleKeyDown(e) {
      if (e.key === key && shortcutEnabled && withContinue) {
        taskComplete();
      }
    }
    window.addEventListener("keydown", handleKeyDown);
    return () => window.removeEventListener("keydown", handleKeyDown);
  });

  return (
    <CenteredNicePaper centerX={centerX} centerY={centerY}>
      <Markdown>{content}</Markdown>
      <br />
      {withContinue && (
        <Button color="primary" onClick={taskComplete}>
          Continue {shortcutEnabled && `(${key})`}
        </Button>
      )}
    </CenteredNicePaper>
  );
};

InformationScreen.propTypes = {
  centerX: PropTypes.bool,
  centerY: PropTypes.bool,
  content: PropTypes.string.isRequired,
  taskComplete: PropTypes.func,
  withContinue: PropTypes.bool,
  shortcutEnabled: PropTypes.bool,
  /* The KeyboardEvent.key value to use as a shortcut */
  key: PropTypes.string,
};

export default InformationScreen;
