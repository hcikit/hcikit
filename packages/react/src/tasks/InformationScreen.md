```js
<InformationScreen
  content={`
# Hello World

Information screens can be written with markdown.
`}
/>
```
