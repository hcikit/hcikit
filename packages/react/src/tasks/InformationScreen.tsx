import React, { useEffect } from "react";
import Markdown from "markdown-to-jsx";

import { Button } from "@material-ui/core";
import { CenteredNicePaper } from "../components";
import PropTypes from "prop-types";
import { useExperiment } from "../core/Experiment";

/**
 * Creates a screen with information or instructions and a continue button.
 * Accepts Markdown.
 * By setting withContinue to false it can also be used as an end screen of an experiment.
 * Uses a keyboard shortcut when specified.
 */
const InformationScreen: React.FunctionComponent<{
  content: string;
  withContinue?: boolean;
  centerX?: boolean;
  centerY?: boolean;
  shortcutEnabled?: boolean;
  shortcut?: string;
}> = ({
  content,
  withContinue = true,
  centerX,
  centerY,
  shortcutEnabled,
  shortcut = "Enter",
}) => {
  const experiment = useExperiment();

  useEffect(() => {
    function handleKeyDown(e: KeyboardEvent) {
      if (e.key === shortcut && shortcutEnabled && withContinue) {
        experiment.taskComplete();
      }
    }
    window.addEventListener("keydown", handleKeyDown);
    return () => window.removeEventListener("keydown", handleKeyDown);
  }, [experiment, shortcut, shortcutEnabled, withContinue]);

  return (
    <CenteredNicePaper centerX={centerX} centerY={centerY}>
      <Markdown>{content}</Markdown>
      <br />
      {withContinue && (
        <Button color="primary" onClick={experiment.taskComplete}>
          Continue {shortcutEnabled && `(${shortcut})`}
        </Button>
      )}
    </CenteredNicePaper>
  );
};

InformationScreen.propTypes = {
  centerX: PropTypes.bool,
  centerY: PropTypes.bool,
  content: PropTypes.string.isRequired,
  withContinue: PropTypes.bool,
  shortcutEnabled: PropTypes.bool,
  /* The KeyboardEvent.key value to use as a shortcut */
  shortcut: PropTypes.string,
};

export default InformationScreen;
