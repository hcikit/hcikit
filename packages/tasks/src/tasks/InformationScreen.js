import React, { useEffect } from "react";
import Markdown from "markdown-to-jsx";

import { Typography, Button } from "@material-ui/core";
import { CenteredNicePaper } from "../components";
import PropTypes from "prop-types";
import { withGridItem } from "../withGridItem";

/**
 * Creates a screen with information or instructions and a continue button.
 * Accepts Markdown.
 * By setting withContinue to false it can also be used as an end screen of an experiment.
 * Uses a keyboard shortcut when specified.
 */
const InformationScreen = ({
  content,
  withContinue = true,
  onAdvanceWorkflow,
  centerX,
  centerY,
  shortcutEnabled,
  key = "Enter"
}) => {
  useEffect(() => {
    function handleKeyDown(e) {
      if (e.key === key && shortcutEnabled) {
        onAdvanceWorkflow();
      }
    }
    window.addEventListener("keydown", handleKeyDown);
    return () => window.removeEventListener("keydown", handleKeyDown);
  });

  return (
    <CenteredNicePaper centerX={centerX} centerY={centerY}>
      <Typography>
        <Markdown children={content} />{" "}
      </Typography>
      <br />
      {withContinue && (
        <Button color="primary" onClick={onAdvanceWorkflow}>
          continue {shortcutEnabled && `(${key})`}
        </Button>
      )}
    </CenteredNicePaper>
  );
};

InformationScreen.propTypes = {
  centerX: PropTypes.bool,
  centerY: PropTypes.bool,
  content: PropTypes.string,
  onAdvanceWorkflow: PropTypes.func,
  withContinue: PropTypes.bool,
  shortcutEnabled: PropTypes.bool,
  /* The KeyboardEvent.key value to use as a shortcut */
  key: PropTypes.string
};

export default withGridItem(InformationScreen);
