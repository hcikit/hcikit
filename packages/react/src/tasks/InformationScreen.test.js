import React from "react";
import { mount } from "enzyme";
import InformationScreen from "./InformationScreen";
import { create, act } from "react-test-renderer";

describe("InformationScreen", () => {
  it("renders without crashing", () => {
    mount(<InformationScreen content="hello" />);
  });

  it("renders without continue", () => {
    let wrapper = mount(
      <InformationScreen withContinue={false} content="hello" />
    );
    expect(wrapper.find("button").exists()).toBeFalsy();
  });

  it("advances to the next task", () => {
    let taskCompleteSpy = jest.fn();
    let container = mount(
      <InformationScreen taskComplete={taskCompleteSpy} content="Hello World" />
    );

    container.find("button").simulate("click");
    expect(taskCompleteSpy).toBeCalledTimes(1);
  });

  it("doesn't advance when with continue is false", () => {
    let taskCompleteSpy = jest.fn();
    const map = {};

    jest.spyOn(window, "addEventListener").mockImplementation((event, cb) => {
      map[event] = cb;
    });

    mount(
      <InformationScreen
        withContinue={false}
        taskComplete={taskCompleteSpy}
        content="Hello World"
      />
    );
    map.keydown({ key: "Enter" });
    expect(taskCompleteSpy).not.toHaveBeenCalled();
  });

  it("advances with a keyboard press", () => {
    const map = {};

    jest.spyOn(window, "addEventListener").mockImplementation((event, cb) => {
      map[event] = cb;
    });

    let taskCompleteSpy = jest.fn();
    mount(
      <InformationScreen
        taskComplete={taskCompleteSpy}
        shortcutEnabled
        content="Hello World"
      />
    );

    map.keydown({ key: "Enter" });
    expect(taskCompleteSpy).toBeCalledTimes(1);
  });

  it("renders markdown properly", () => {
    let markdown = `
# Hello World

 - This is some markdown
 - that should render *properly*
`;

    let root;

    act(() => {
      root = create(<InformationScreen content={markdown} />);
    });

    expect(root.toJSON()).toMatchSnapshot();
  });
});
