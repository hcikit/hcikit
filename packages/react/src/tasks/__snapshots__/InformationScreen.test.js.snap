// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InformationScreen renders markdown properly 1`] = `
<div
  style={
    Object {
      "gridArea": "task",
    }
  }
>
  <div
    className="sc-ifAKCX iCjAbu"
  >
    <div>
      <div
        className="MuiPaper-root MuiPaper-elevation1 sc-bxivhb hoTgWR MuiPaper-rounded"
      >
        <div>
          <h1
            id="hello-world"
          >
            Hello World
          </h1>
          <ul>
            <li>
              This is some markdown
            </li>
            <li>
              that should render 
              <em>
                properly
              </em>
            </li>
          </ul>
        </div>
        <br />
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
          disabled={false}
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            continue 
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
